### DISCUSSION TITLE
How to configure DMS for testing as an SMTP server locally?

### DISCUSSION THREAD

**Discussion by jaimebw** [docker-mailserver/docker-mailserver#3635] [Q&A] [answered]
Hi, I'm writing this issue just to ask for help for setting up `docker-mailserver`.

Currently, I'm trying to develop some stuff and I need to run `docker-mailserver` on local (mainly to run tests). The IMAP server works fine but the SMTP server is another story...

I'm using `smtplib` in Python. This an example of script of what I am using to test the SMTP server:

```python
import imaplib
server = smtplib.SMTP("localhost", 25)
server.login(f"test@mail.local", "test")
```

The error I'm running into looks like this:

```
>               raise SMTPServerDisconnected("Connection unexpectedly closed")
E               smtplib.SMTPServerDisconnected: Connection unexpectedly closed

/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/smtplib.py:405: SMTPServerDisconnected
```

My `compose.yml` looks like this:

```yml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.mail.local
    env_file: local-mailserver.env
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/latest/config/security/understanding-the-ports/
    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    ports:
      - "25:25"
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
      - "5000:5000" # Localhost
    volumes:
      - ./docker-data/certbot/certs/:/etc/letsencrypt
      - ./docker-data/dms/mail-data/:/var/mail/
      - ./docker-data/dms/mail-state/:/var/mail-state/
      - ./docker-data/dms/mail-logs/:/var/log/mail/
      - ./docker-data/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
    restart: always
    stop_grace_period: 1m
    # Uncomment if using `ENABLE_FAIL2BAN=1`:
    # cap_add:
    #   - NET_ADMIN
    healthcheck:
      test: "ss --listening --tcp | grep -P 'LISTEN.+:smtp' || exit 1"
      timeout: 3s
```

The `local-mailserver.env` is just the default `mailserver.env` provided from the repo (no SSL certificates)

As stated before, I have no problems connecting to the IMAP server.

If you need more details,  please let me know. Thanks in advance!

**Comment by polarathene** [comment, marked as answer]
> My `compose.yml` looks like this

What is the `5000:5000` port binded for?

Why do you need `stop_grace_period` or `healthcheck` in this local testing scenario? Or many of the volume mounts, especially the `letsencrypt` seems pointless?

`hostname: mail.mail.local` is a bad choice, **`.local` is for mDNS**, use a more appropriate TLD like `mail.example.test`. `.test` is reserved for this purpose, no public DNS resolver concerns to worry about and no potential mDNS resolving issues from `.local`.

~~You also reference an ENV file for some reason, is there anything note worthy there? If not don't use that and keep your config simple by explicitly having ENV in your `compose.yaml`, otherwise when sharing your config you should also share this file so we know what ENV you're using.~~ (_**EDIT:** user clarified they're just using the default file from our repo_)

---

> This an example of script of what I am using to test the SMTP server
> 
> ```python
> server = smtplib.SMTP("localhost", 25)
> server.login(f"test@mail.local", "test")
> ```

You are connecting to port 25 which is for delivering mail from another MTA (mail server).

You want to bypass security checks involved there and instead submit mail to the mail-server directly, where it will either be delivered internally (_to a configured mail account_), or sent out from DMS to another mail server (_potentially another DMS container instance_).

To do this you need to use authenticated submission(s) ports with credentials which is what your `server.login()` looks to be intended for:
- Port 465 requires implicit TLS, while port 587 is explicit TLS via StartTLS (_usually this allows plaintext fallback, but it is not permitted for authentication due to security reasons, requiring TLS cert_)
- [We do not support login over port 25](https://github.com/docker-mailserver/docker-mailserver/pull/3006).
- For testing purposes, you can [avoid setting `SSL_TYPE`](https://docker-mailserver.github.io/docker-mailserver/latest/config/environment/#ssl_type) and this will [disable the security requirement on ports 465 and 587](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/scripts/helpers/ssl.sh#L316-L360). Alternatively, there presently is an [undocumented `snakeoil` option](https://github.com/docker-mailserver/docker-mailserver/blob/26214491efb2ec07242c24a9191633f0d72998d9/target/scripts/helpers/ssl.sh#L362-L365), it is used for testing TLS connections when the validity of the certificate doesn't actually matter, in future it won't be available and the alternative empty / unset value for disabling will be changed to `disabled`.
- You may also generate your own local testing certificate yourself if you need that. I have [documented this](https://github.com/docker-mailserver/docker-mailserver/tree/26214491efb2ec07242c24a9191633f0d72998d9/test/test-files/ssl/example.test) (_and a similar reference in our web docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/security/ssl/#generating-a-self-signed-certificate)_) and if you use the same `mail.example.test` or `example.test` FQDN for your testing container, you can use the cert files provided there directly.

As a reminder, `mail.example.test` is the FQDN for DMS as the mailserver.
- The default assumption will be mail accounts use `@example.test` but you could have accounts with any domain-part (_the part of email address after `@`_), such as `@another-example.test`, this has nothing to do with the FQDN for DMS, DNS will have the domain-part lookup the MX record for delivery (_that points to `mail.example.test`_).
- There's also another common misunderstanding users run into. A TLS cert should be provisioned for the FQDN for DMS, not the domain-part of mail accounts it manages.

---

## TL;DR

- Use `hostname: mail.example.test`, this is a better choice with the `.test` TLD than `.local` (intended for mDNS).
- Use port 465 or 587 (StartTLS, like port 25). These are submission ports that require authentication via login, port 25 is only for untrusted third-party deliveries.
- If you do not provide local testing certificates, ensure that `SSL_TYPE=''` (_empty / unset_). This will presently disable requiring a secure connection.

Then the following will work:

```python
server = smtplib.SMTP("localhost", 465)
server.login(f"test@example.test", "hunter42")
```

**Reply by jaimebw** [reply]
Thank you very much! The issue was this + the fact that my system is Mac Os( I should have said this on my initial question -_-). Found this [answer](https://github.com/docker-mailserver/docker-mailserver/issues/3530), did the same and I finally I am able to connect to the SMTP server. 
Thanks again, I really apreciate the effort on your response.
